### PR TITLE
core/upgrade/: Add `ReadyUpgrade`

### PR DESCRIPTION
**Pull request by thomaseizinger**
# Description

This is the counterpart to `PendingUpgrade`.

## Links to any relevant issues

Extracted out of https://github.com/libp2p/rust-libp2p/pull/2852.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
~- [ ] I have added tests that prove my fix is effective or that my feature works~
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+# 0.35.2 [unreleased]
+
+- Add `ReadyUpgrade`. See [PR XXXX].
+
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline core/CHANGELOG.md]:
```suggestion
- Add `ReadyUpgrade`. See [PR 2855].

[PR 2855]: https://github.com/libp2p/rust-libp2p/pull/2855
```

**Comment by thomaseizinger** [comment]
I am planning to use this throughout the codebase where applicable. Should I do this as part of this PR or send separate ones @mxinden? It could for example be used in the ping protocol.

**Review by mxinden — COMMENTED** [review]
I am in favor of introducing a `ReadyUpgrade`.

Note that long term I would like to get rid of the upgrade mechanism in `ConnectionHandler`, though that still needs more thought.

> I am planning to use this throughout the codebase where applicable. Should I do this as part of this PR or send separate ones @mxinden? It could for example be used in the ping protocol.

I think every pull request should be:

- As small as possible.
- Atomic.
- Leave `master` in the same or better state.

For the sake of the latter, concretely here for the sake of code deduplication, I think this pull request should include the changes across the many use-cases.

**Comment by thomaseizinger** [comment]
@mxinden The only place where it was directly applicable was the ping protocol.

File: protocols/ping/src/protocol.rs [hunk 2]
```diff
@@ -52,35 +49,6 @@ pub struct Ping;
...
-impl OutboundUpgrade<NegotiatedSubstream> for Ping {
-    type Output = NegotiatedSubstream;
-    type Error = Void;
-    type Future = future::Ready<Result<Self::Output, Self::Error>>;
-
-    fn upgrade_outbound(self, stream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
-        future::ok(stream)
-    }
-}
-
```
Comment (mxinden) [inline protocols/ping/src/protocol.rs:83]: _Emphasizing this, though I think we are of the same opinion:_ While not de-duplicating given that `ReadyUpgrade` is only used once, I do think these kind of pull requests do a great job simplifying the business logic, here `libp2p-ping`, moving auxiliary not-directly-related logic to common places, here `libp2p-core`.

File: protocols/ping/src/handler.rs [hunk 3]
```diff
@@ -18,10 +18,11 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use crate::protocol;
+use crate::{protocol, PROTOCOL_NAME};
 use futures::future::BoxFuture;
 use futures::prelude::*;
 use futures_timer::Delay;
+use libp2p_core::upgrade::ReadyUpgrade;
```
Comment (mxinden) [inline protocols/ping/src/handler.rs:25]: Given that `libp2p-ping` now depends on `libp2p-core` `v0.35.2`, the `libp2p-ping` `Cargo.toml` needs to be adjusted, right?

**Review by mxinden — COMMENTED** [review]
Thanks for the follow up!

File: protocols/ping/src/protocol.rs (hunk 2 above)
Comment (thomaseizinger) [reply protocols/ping/src/protocol.rs:83]: Yes, strong agreement here.

File: protocols/ping/src/handler.rs (hunk 3 above)
Comment (thomaseizinger) [reply protocols/ping/src/handler.rs:25]: Good catch!

**Comment by thomaseizinger** [comment]
Comments addressed, ready to merge from my end @mxinden!

**Review by mxinden — APPROVED** [review]

**Pull request merged**